nifi-nar-bundles/nifi-oraclecdc-bundle/nifi-oraclecdc-processors/src/main/java/org/apache/nifi/processors/oraclecdc/controller/impl/StandardOracleCDCService.java

```diff
@@ -0,0 +1,269 @@
+            Class<?> xstreamOut = loadClass("oracle.streams.XStreamOut");
+            Method method = xstreamOut.getMethod("receiveLCRCallback",
+                    loadClass("oracle.streams.XStreamLCRCallbackHandler"), int.class);
+            method.invoke(xsOut, proxy,
+                    loadClass("oracle.streams.XStreamOut").getDeclaredField("DEFAULT_MODE").getInt(null));
+            System.out.print("done");
+        } catch (Exception e) {
+            e.printStackTrace();
+            getLogger().warn("exception when processing LCRs");
+            getLogger().warn(e.getMessage());
+            throw new ProcessException("exception when processing LCRs " + e.getMessage());
+        }
+    }
+
+    @Override
+    public void setProcessedLowWaterMark(Object xsOutServer, byte[] position) {
+        try {
+            System.out.println(new String(new Base32(true).encode(position)));
+            Class<?> xstreamOut = loadClass("oracle.streams.XStreamOut");
+            Method method = xstreamOut.getMethod("setProcessedLowWatermark", byte[].class, int.class);
+            method.invoke(xsOutServer, position,
+                    loadClass("oracle.streams.XStreamOut").getDeclaredField("DEFAULT_MODE").getInt(null));
+        } catch (Exception e) {
+            e.printStackTrace();
+            getLogger().warn("exception when set low watermark");
+            getLogger().warn(e.getMessage());
+            throw new ProcessException("exception when setting low water mark " + e.getMessage());
+        }
+    }
+
+    @Override
+    public Object attach(String xsOutName, byte[] lastPosition) {
+        getLogger().info("in attach");
+        try {
+            final Connection conn = getConnection();
+            Class<?> xstreamOut = loadClass("oracle.streams.XStreamOut");
+            Method method = xstreamOut.getMethod("attach", loadClass("oracle.jdbc.OracleConnection"), String.class,
+                    byte[].class, int.class);
+            Object xsOut = method.invoke(null, conn.unwrap(loadClass("oracle.jdbc.OracleConnection")), xsOutName,
+                    lastPosition, xstreamOut.getDeclaredField("DEFAULT_MODE").getInt(null));
+            return xsOut;
+        } catch (Exception e) {
+            System.out.println("cannot attach to outbound server: " + xsOutName);
+            getLogger().error(e.getMessage());
+            e.printStackTrace();
+            throw new ProcessException("cannot attach to outbound server: " + xsOutName);
+        }
+    }
+
+    @Override
+    public void detach(Object xsOut) {
+        try {
+            Class xstreamOut = loadClass("oracle.streams.XStreamOut");
+            Method method = xstreamOut.getDeclaredMethod("detach", int.class);
+            method.invoke(xsOut, xstreamOut.getDeclaredField("DEFAULT_MODE").getInt(null));
+        } catch (Exception e) {
+            getLogger().info("cannot detach from the outbound server: ");
```

Comment:
Probabye `e` should be added as second paramater for both `getLogger().info` and `new ProcessException`.

Reply:
agreed, did the change

getLogger().warn("cannot detach from the outbound server: "+e.getMessage());
            throw new ProcessException("cannot detach from the outbound server: "+e.getMessage());